MdeModulePkg/Core/Dxe/Misc/MemoryProtectionSupport.c

```diff
@@ -2792,6 +2787,120 @@ GetDxeMemoryProtectionSettings (
   return NULL;
 }
 
+/**
+  Enable NULL pointer detection by changing the attributes of page 0. The assumption is that PEI
+  has set page zero to allocated so this operation can be done safely.
+
+  @retval EFI_SUCCESS       Page zero successfully marked as read protected
+  @retval Other             Page zero could not be marked as read protected
+
+**/
+VOID
+EFIAPI
+EnableNullDetection (
+  VOID
+  )
+{
+  EFI_STATUS                       Status;
+  EFI_GCD_MEMORY_SPACE_DESCRIPTOR  Desc;
+
+  Status = CoreGetMemorySpaceDescriptor (0, &Desc);
+
+  if (EFI_ERROR (Status)) {
+    return;
+  }
+
+  if ((Desc.Capabilities & EFI_MEMORY_RP) == 0) {
+    Status = CoreSetMemorySpaceCapabilities (
+               0,
+               EFI_PAGES_TO_SIZE (1),
+               Desc.Capabilities | EFI_MEMORY_RP
+               );
+    if (EFI_ERROR (Status)) {
+      ASSERT_EFI_ERROR (Status);
+      return;
+    }
+  }
+
+  Status = CoreSetMemorySpaceAttributes (
+             0,
+             EFI_PAGES_TO_SIZE (1),
+             Desc.Attributes | EFI_MEMORY_RP
+             );
+  ASSERT_EFI_ERROR (Status);
+}
+
+/**
+  Disable NULL pointer detection.
+**/
+VOID
+DisableNullDetection (
+  VOID
+  )
+{
+  EFI_STATUS                       Status;
+  EFI_GCD_MEMORY_SPACE_DESCRIPTOR  Desc;
+
+  DEBUG ((DEBUG_INFO, "%a - Enter\n", __FUNCTION__));
+
+  Status = CoreGetMemorySpaceDescriptor (0, &Desc);
+
+  if (EFI_ERROR (Status)) {
+    DEBUG ((
+      DEBUG_ERROR,
+      "%a - Failed to get memory space descriptor for NULL address! Status = %r\n",
+      __FUNCTION__,
+      Status
+      ));
+    return;
+  }
+
+  if ((Desc.Attributes & EFI_MEMORY_RP) == 0) {
+    DEBUG ((DEBUG_INFO, "%a - Nothing to do... Exit\n", __FUNCTION__));
+    return;
+  }
+
+  if ((Desc.Capabilities & EFI_MEMORY_RP) == 0) {
+    Status = CoreSetMemorySpaceCapabilities (

```

Comment:
Why do we set the capability here? Does the set attribute call fail to remove this attribute if the capability is not set?

Reply:
I wrote this to mirror the upstream DisableNullDetectionAtTheEndOfDxe() function but looking at the ConvertSpace() logic it doesn't seem like it's necessary to set the capabilities since this is clearing the attribute.

Reply:
Removed